NEXT-7754 - Fix thumbnail generation when physical file is missing

leptoquark1 opened the pull request:
### 1. Why is this change necessary?

Regeneration of thumbnails requires a check of the physical file to ensure that the thumbnail was actually generated and not just having a virtual representation in the database.

The circumstances why this situation might occur are not relevant.

However, possible causes could be:

 - Third-party actions such as file monitoring processes and automated clean-up operations
 - Unintentional deletion by humans
 - Previously unknown Bugs in Code that can cause such a condition
 - ...

### 2. What does this change do, exactly?

Adds a physical file check when updating thumbnails.


### 3. Describe each step to reproduce the issue or behaviour.

- Make sure you have generated thumbnails.
- Delete the Thumbnail folder or only a specific file from the sub-folders
- Execute `bin/console media:generate-thumbnails`
- Check that every thumbnail generation was skipped

### 4. Please link to the relevant issues (if any).

- [NEXT-7754](https://issues.shopware.com/issues/NEXT-7754)
- [artikelbild-thumbnails-verschwinden-sind-kaputt-404-wieso](https://forum.shopware.com/t/artikelbild-thumbnails-verschwinden-sind-kaputt-404-wieso/69272)

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

Reopens #1879 @J-Rahe

JoshuaBehrens commented:
@leptoquark1 is back 🎉  Could've used that one right yesterday :D

J-Rahe commented:
Hey @leptoquark1, there was an discussion regarding the performance of this yesterday. I will come back to you once i have an answer on how to proceed.

7underlines commented:
What are the performance concerns here?
I would love to see this merged.

jakob-kruse commented:
Hey @leptoquark1 and @treffner, sorry for the late response.

We decide to implement this check as an optional flag E.g. "--strict" to the media:generate-thumbnails command, due to the performance impact.

Note that adding an optional parameter to `updateThumbnails` is [breaking](https://developer.shopware.com/docs/resources/guidelines/code/backward-compatibility)

leptoquark1 commented:
@jakob-kruse I will implement the missing flag to the command and its further behaviour in the service in the next few days.

Commit:
Refactor GenerateThumbnailsHandler to distinguish between message types

Depending on whether you add the flag `async` to the 'media:generate-thumbnails' or not, it will
 run either the `generate` or `updateThumbnails` method of `ThumbnailService` class,
  which may lead to different behaviour or results.

The `GenerateThumbnailsCommand` calls the ThumbnailService::updateThumbnails when running synchronous,
 but when `async` flag is provided, an `UpdateThumbnailsMessage` will be dispatched without the
 provided parameter as in ref a49510e2 introduced.

The responsible handler (`GenerateThumbnailsHandler`) handles both, the `GenerateThumbnailsMessage`
 and `UpdateThumbnailsMessage`. However, it does proceed the same for both types and calls method
 `ThumbnailService::generate`,ignoring the specific type of message.

With this change, the `GenerateThumbnailsHandler` distinguishes which method is executed in the
 `ThumbnailService` based on the message type.  Furthermore, the new flag `strict` (ref a49510e2)
 is passed on as desired and considered in the thumbnail generation process as expected.

Commit:
NEXT-7754 - Fix thumbnail generation when physical file is missing

Fix for the method `updateThumbnails` function in `src/Core/Content/Media/Thumbnail/ThumbnailService.php` so that the physical presence of the thumbnail file is respected. This is handled by new parameter `$strict`.
To use the new behaviour with the `media:generate-thumbnails` command, a new option `--strict` is added. When supplied, it will be passed to the fixed method to be apply the fixed behaviour.

leptoquark1 commented:
@jakob-kruse I've updated the commit as requested.

But I was not sure about the  `media:generate-thumbnails` command behaviour with `--async` option.

It does dispatch the `UpdateThumbnailsMessage`, but the corresponding `GenerateThumbnailsHandler` calls the method `generate` of the `ThumbnailService`.

By this step I noticed that the `updateThumbnails` and `generate` method does not share any accumulation logic until the call of `createThumbnailsForSizes`.
This may cause a different behaviour of the thumbnails being update/generated depending on if the `async` option is given in with `media:generate-thumbnails` or not. I was not sure if this was intended.

However, I only implemented the target call of updateThumbnails, but I'd rather suggest to refactor the `ThumbnailService` or add an condition in the `GenerateThumbnailsHandler`:


```php
class GenerateThumbnailsHandler extends AbstractMessageHandler
{
    // [...]

    public function handle($msg): void
    {
        // [...]

        if ($msg instanceof UpdateThumbnailsMessage) {
            foreach ($entities as $media) {
                $this->thumbnailService->updateThumbnails($media, $context, $msg->isStrict());
            }
        } else {
            $this->thumbnailService->generate($entities, $context);
        }
    }

    // [...]
}
```

```php
class UpdateThumbnailsMessage extends GenerateThumbnailsMessage
{
    private bool $isStrict = false;
     
    public function isStrict(): bool
    {
        return $this->isStrict;
    }

    public function setIsStrict(bool $isStrict): void
    {
        $this->isStrict = $isStrict;
    }
}
```

```php
class GenerateThumbnailsCommand extends Command
{
    // [...]

    private function generateAsynchronous(RepositoryIterator $mediaIterator, Context $context): void
    {
        // [...]
        while (($result = $mediaIterator->fetch()) !== null) {
            $msg = new UpdateThumbnailsMessage();
            // [...]
            $msg->setIsStrict($this->isStrict); // <--
            $this->messageBus->dispatch($msg);
            // [...]
        }
        // [...]
    }

    // [...]
}

jakob-kruse commented:
Hey @leptoquark1 

Thanks for adding the option. I noticed there were some changes to the theme compiling in the meantime. It would be nice if you could rebase your changes.

I will get back to you once I know more!

leptoquark1 commented:
> Hey @leptoquark1
> 
> Thanks for adding the option. I noticed there were some changes to the theme compiling in the meantime. It would be nice if you could rebase your changes.
> 
> I will get back to you once I know more!

@jakob-kruse Due to a merge conflict with current trunk, I have rebased this pull request again.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/platform?pullRequest=1937) <br/>All committers have signed the CLA.

fschmtt commented:
Hi @leptoquark1,

in this https://github.com/shopware/platform/pull/1937#issuecomment-908302468 you also suggested changes for the case when thumbnails will be generated asynchronously.

If you just waited for feedback on that, I think it's fine and you can refactor it.

leptoquark1 commented:
@fschmtt @keulinho @jakob-kruse Sorry for the late response!

Looks like I just dismissed the notification and forgot about it... However, I made the changes as I suggested and it works just fine.

This pull request now contains the following changes, which are split into two commits:

* Fix `updateThumbnails` function in `src/Core/Content/Media/Thumbnail/ThumbnailService.php` so that the physical presence of the thumbnail file is respected
* Added new parameter `$strict` to method `Shopware\Core\Content\Media\Thumbnail\ThumbnailService::updateThumbnails`
* Added option `--strict` to command `media:generate-thumbnails`. Supplying it, a physical file check for existing thumbnails is performed.


* Fix different behaviour or results expected when `media:generate-thumbnails` command is executed with or without the `async` flag.
* The `GenerateThumbnailsHandler` now distinguish which method is executed in the`ThumbnailService` based on the message type.
* Added two methods `isStrict` and `setIsStrict` and it's corresponding property `isStrict` to class `UpdateThumbnailsMessage`
* The `GenerateThumbnailsHandler` will now proceed with `ThumbnailService::generate` when handling `GenerateThumbnailsMessage`
* The `GenerateThumbnailsHandler` will proceed with `ThumbnailService::updateThumbnails` when handling `UpdateThumbnailsMessage`
* When processed by MessageBus the optional strict flag for thumbnail generation (as introduced in _NEXT-7754_) will be passed,
  to the respective `ThumbnailService::updateThumbnails` method as expected. So it may considered in following generation process.


I would actually recommend combining the two methods `generate` and `updateThumbnails` to create some sort of "single source of trouth", since it should be no difference in its context and responsibility if the input is either update or a generate. **But we should close this issue and have these changes solve existing issues.**

Have a good day :)

fschmtt commented:
Wow, many thanks for your efforts @leptoquark1! 🎉

keulinho commented:
Hey @leptoquark1 thanks for your contribution! :blue_heart: :tada:

leptoquark1 commented:
#383 
#1062
